pkg/agent/openflow/client.go

```diff
@@ -1128,10 +1125,7 @@ func (c *client) SendICMPPacketOut(
 	packetOutBuilder = packetOutBuilder.SetICMPCode(icmpCode)
 	packetOutBuilder = packetOutBuilder.SetICMPData(icmpData)
 
-	// Reject response packet should bypass ConnTrack
-	if isReject {
-		packetOutBuilder = packetOutBuilder.AddLoadRegMark(CustomReasonRejectRegMark)
-	}
+	packetOutBuilder = addActionsToPacketOut(packetOutBuilder, packetOutType)
```

Comment:
Is it possible to create a new function dedicated for "reject service traffic"? And you can call `addActionsToPacketOut` once in this new function, then append the "AddResubmitAction" in any flowbuilder which needs it. In this way, you don't need to call `addActionsToPacketOut` twice, and you don't need to inject this specific function in the general function "SendICMPPacketOut" and "SendTCPPacketOut" which don't require this specific action in other cases.

Reply:
I do agree with Wenying that the current code is quite clunky. However, I don't think that these functions (`SendTCPPacketOut` / `SendICMPPacketOut`) should be responsible for doing anything which is specific to reject packets. I do not think that having `binding.PacketOutType` (which is specific to reject packets) as a parameter is a good idea.

Instead I think the last argument to these functions should be an arbitrary "PacketOutBuilder mutator":
```go
func (c *client) SendICMPPacketOut(
	srcMAC string,
	dstMAC string,
	srcIP string,
	dstIP string,
	inPort uint32,
	outPort uint32,
	isIPv6 bool,
	icmpType uint8,
	icmpCode uint8,
	icmpData []byte,
	mutatePacket func(builder PacketOutBuilder) PacketOutBuilder) error {
```

Reply:
Thanks for the review. An arbitrary mutator func sounds good to me. I understand what you mean and at the same time, I realized that I didn't express my thoughts clearly. Sorry for the confusion. I'd like to explain a little bit. Then you guys can see if it still doesn't make sense.
I don't want to make SendTCPPacketOut or SendICMPPacketOut specific to reject packet. I'd like to make PacketOutType as a generic type. SendTCPPacketOut or SendICMPPacketOut could add OF Actions to packetOutBuilder based on PacketOutType.
Currently, I only included cases for reject, so I name the type with a generic name PacketOutType and all current types have the prefix Reject. If needed, we could add more packetOutTypes in the future. And at the same time, add OF Actions that new packetOutType need in func addActionsToPacketOut.
Maybe add a normal type PacketOutType which won't do anything in func addActionsToPacketOut could help others understand?
Let me know if this makes sense to you guys. Thanks!

Reply:
I think you have too many cases specific to the reject case (7) to make it a good candidate for a generic packet type concept. It would be a bad idea to have such a detailed list of cases that just keeps on growing. It's better to have a specialized "reject packet type" concept, and keep it local to the code dealing with reject packets.

Then you can have a `SendRejectICMP` function which is going to call `c.SendICMPPacketOut` with the right parameters (ICMP type / code, packet mutator).